dumper/app/src/main/java/com/google/edwmigration/dumper/application/dumper/task/AbstractJdbcTask.java

```diff
@@ -76,28 +83,81 @@ public AbstractJdbcTask<T> withHeaderClass(@Nonnull Class<? extends Enum<?>> hea
     CSVFormat format = FORMAT;
     Class<? extends Enum<?>> headerClass = getHeaderClass();
     if (headerClass != null) {
       format = format.withHeader(headerClass);
-      if (headerClass.getEnumConstants().length != rs.getMetaData().getColumnCount())
-        // Can we avoid nesting exceptions here?
-        throw new SQLException(
-            new MetadataDumperUsageException(
-                "Fatal Error. ResultSet does not have the expected column count: "
-                    + headerClass.getEnumConstants().length,
-                Arrays.asList(
-                    "If a custom query has been specified please confirm the selected columns match"
-                        + " the following: ",
-                    StringUtils.join(headerClass.getEnumConstants(), ", "))));
+      ResultSetMetaData metaData = rs.getMetaData();
+      validateHeader(headerClass, metaData);
     } else {
       format = format.withHeader(rs);
     }
     return format;
   }
 
+  private void validateHeader(Class<? extends Enum<?>> headerClass, ResultSetMetaData metaData)
+      throws SQLException {
+    if (!requiresHeaderValidation) return;
+
+    if (ignoreExtraColumns) {
+      Set<String> resultColumnNames = new HashSet<>();
+
+      for (int i = 0; i < metaData.getColumnCount(); i++) {
+        resultColumnNames.add(metaData.getColumnName(i + 1));
+      }
+
+      Set<String> remainingColumnNames = calculateRemainingColumns(headerClass, resultColumnNames);
+
+      if (!remainingColumnNames.isEmpty()) {
+        throw new SQLException(
+            "Fatal Error. ResultSet does not have the expected columns: " + remainingColumnNames);
+      }
+    } else if (headerClass.getEnumConstants().length != metaData.getColumnCount())
+      throw new SQLException(
+          "Fatal Error. ResultSet does not have the expected column count: "
+              + headerClass.getEnumConstants().length
+              + "\n"
+              + "If a custom query has been specified please confirm the selected columns match"
+              + " the following: "
+              + StringUtils.join(headerClass.getEnumConstants(), ", "));
+  }
+
+  static Set<String> calculateRemainingColumns(
+      Class<? extends Enum<?>> headerClass, Set<String> resultColumnNames) {
+    Set<String> availableCamelCaseNames =
+        resultColumnNames.stream()
+            .map(name -> CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, name))
+            .collect(Collectors.toSet());
+    Set<String> availableUnderscoreCaseNames =
+        resultColumnNames.stream()
+            .map(name -> CaseFormat.UPPER_UNDERSCORE.to(CaseFormat.LOWER_UNDERSCORE, name))
+            .collect(Collectors.toSet());
+
+    Set<String> requiredNames =
+        Arrays.stream(headerClass.getEnumConstants())
+            .map(enumVal -> CaseFormat.UPPER_CAMEL.to(CaseFormat.LOWER_UNDERSCORE, enumVal.name()))
+            .collect(Collectors.toSet());
```

Comment:
Why are we manipulating the names? Is there a mismatch between some TD versions? Does that manipulation extend to underscores? Or is this to do with our convention to have enums all have UPPER_UNDERSCORE, which doesn't match the underlying?